Add content type intro and FTI reference (#1498)
plone/documentation: docs/backend/content-types/index.md

         to="@@edit"
  />
  <alias from="sharing"
         to="@@sharing"
  />
  <alias from="view"
         to="(selected layout)"
  />

  <!-- Actions -->
  <action action_id="view"
          category="object"
          condition_expr=""
          icon_expr="string:toolbar-action/view"
          title="View"
          url_expr="string:${object_url}"
          visible="True"
          i18n:attributes="title"
  >
    <permission value="View" />
  </action>
  <action action_id="edit"
          category="object"
          condition_expr="not:object/@@plone_lock_info/is_locked_for_current_user|python:True"
          icon_expr="string:toolbar-action/edit"
          title="Edit"
          url_expr="string:${object_url}/edit"
          visible="True"
          i18n:attributes="title"
  >
    <permission value="Modify portal content" />
  </action>

</object>
```

The `name` attribute on the root element in the XML must match the name in the filename and the name listed in `types.xml`.

Set the `i18n:domain` to the i18n domain which includes translations for this content type. This is usually the same as the name of the Python package which contains the content type.

> The XML sets a number of FTI properties:

Reviewer: What's the difference between this list and the next? I think it is meant to be global, correct?

Perhaps a subheading for each list would help clarify. As a bonus, it would provide a visual reference and permalink. When I look at the preview, it visually blurs together as one big list.

```suggestion

(global-fti-properties-label)=

### Global FTI properties

The XML sets a number of FTI properties that are used globally, in both Classic UI and Volto:
```

Author: @stevepiercy good idea

:   Name of the factory adapter used to create new instances of the content type.
    Usually the same as the content type name.

`filter_content_types`
:   Boolean.
    Controls which content types can be added inside this one.
    If `True`, allow only the types listed in `allowed_content_types`.
    If `False`, allow any content type that the user has permission to add.

`global_allow`
:   Boolean.
    Set to `True` to allow adding the content type anywhere in the site where the user has permission.
    Set to `False` to only allow adding it inside other content types that include this one in `allowed_content_types`.

`klass`
:   Dotted path to the Python class for this content type.

`model_file`
:   Location of an XML file to load as the content type's schema.
    This is an alternative to `schema` and `model_source`.

`model_source`
:   Inline XML schema for the content type.
    This is an alternative to `schema` and `model_file`.

`schema`
:   Dotted path to the Python schema for this content type.
    One of `model_file`, `model_source`, and `schema` must be set.
    `schema` is the most commonly used.

`title`
:   The name of the content type displayed in the UI.

> The following FTI properties are only used in the Classic UI:

Reviewer: ```suggestion

(classic-ui-only-fti-properties-label)=

### Classic UI only FTI properties

The following FTI properties are used only in Classic UI:
```

`add_view_expr`

> :  {term}`TALES` expression returning the URL for the form to add a new item of this content type.

Reviewer: Make indent consistent.

```suggestion
:   {term}`TALES` expression returning the URL for the form to add a new item of this content type.
```

# Content Types

```{seealso}
See the chapter {ref}`training:dexterity1-label` from the Mastering Plone 6 Training for a step-by-step tutorial to create a custom content type.
```

## What is a content type?

> Each item in a Plone site is an instance of a particular content type. We have different content types to reflect the different kinds of information about which we need to collect and display information.

Reviewer: One sentence per line, here and elsewhere. I'll just make the one comment for this style guide matter.

```suggestion
Each item in a Plone site is an instance of a particular content type.
We have different content types to reflect the different kinds of information about which we need to collect and display information.
```

> Pages, news items, events, files (binary) and images are examples of content types.

Reviewer: Oxford comma.

```suggestion
Pages, news items, events, files (binary), and images are examples of content types.
```

Lots of things in Plone can be configured to work differently based on the content type. For example, each content type has:
- a {ref}`schema <backend-fields-label>` specifying the fields which can be edited for the content type
- a list of {ref}`behaviors <backend-behaviors-label>` which supply additional functionality that can be attached to the content types for which the behavior is enabled
- a {ref}`workflow <backend-workflows-label>` controlling transitions between publishing states and associated permissions
- a version policy controlling whether to store a revision history

It is common in developing a web site that you'll need customized versions of common content types, or perhaps even entirely new types.

## Factory Type Information

> A content type is defined by creating a Factory Type Information (FTI) object.

Reviewer: Refer to existing Glossary term.

```suggestion
A content type is defined by creating a {term}`Factory Type Information` (FTI) object.
```

To create an FTI in a GenericSetup profile, add the content type to the list in `types.xml`. For example, this adds the standard Plone page (Document) content type:

```xml
<object name="portal_types">
  <object name="Document" meta_type="Dexterity FTI" />
</object>
```

Then, add a file to the `types` directory with the same name.
In this example, the file is `types/Document.xml` and contains this XML:

> ```xml

Reviewer: Insert blank line.

```suggestion

```xml
```

  <property name="default_view_fallback">False</property>
  <property name="immediate_view">view</property>
  <property name="view_methods">
    <element value="document_view" />
  </property>

  <!-- Method aliases -->
  <alias from="(Default)"
         to="(dynamic view)"
  />
  <alias from="edit"
         to="@@edit"
  />
  <alias from="sharing"
         to="@@sharing"
  />
  <alias from="view"
         to="(selected layout)"
  />

  <!-- Actions -->
  <action action_id="view"
          category="object"
          condition_expr=""
          icon_expr="string:toolbar-action/view"
          title="View"
          url_expr="string:${object_url}"
          visible="True"
          i18n:attributes="title"
  >
    <permission value="View" />
  </action>
  <action action_id="edit"
          category="object"
          condition_expr="not:object/@@plone_lock_info/is_locked_for_current_user|python:True"
          icon_expr="string:toolbar-action/edit"
          title="Edit"
          url_expr="string:${object_url}/edit"
          visible="True"
          i18n:attributes="title"
  >
    <permission value="Modify portal content" />
  </action>

</object>

> ```

Reviewer: ```suggestion
```

Author: @stevepiercy I don't understand this suggestion. It's the end of the xml code block that started on line 46.

Reviewer: My mistake. GitHub diff confused me.

> The `name` attribute on the root element in the XML must match the name in the filename and the name listed in `types.xml`.

Reviewer: Add blank line.

```suggestion

The `name` attribute on the root element in the XML must match the name in the filename and the name listed in `types.xml`.
```

Author: @stevepiercy I don't understand the suggestion; line 144 is already blank.

Reviewer: My mistake. GitHub diff confused me.

> The `name` attribute on the root element in the XML must match the name in the filename and the name listed in `types.xml`.

Reviewer 2: Isn't it, that the filename must match as well? Shall we mention it? Here? Above?

Author: The filename is already mentioned in this sentence.

The XML sets a number of FTI properties:

> ```{glossary}

Reviewer: Convert to a [definition list](https://myst-parser.readthedocs.io/en/latest/syntax/typography.html#definition-lists-and-glossaries) to avoid multiple glossaries.

```suggestion
```

Author: Will do, but I found multiple glossaries already in other parts of the documentation so didn't realize it was a problem.

Author: Done.

> :sorted: true

Reviewer: Instead manually sort by the term.

```suggestion
```

> title

Reviewer: Example definition term: make the attribute an inline literal.

```suggestion
`title`
```

>     The name of the content type displayed in the UI.

Reviewer: Leading `:` for the definition. Then repeat for the rest of the definition list items. Only one colon is needed per definition, and subsequent sentences should be indented 4 spaces.

```suggestion
:   The name of the content type displayed in the UI.
```

`add_permission`
:   Id of the permission controlling whether the current user has permission to add this content type.

`allow_discussion`
:   Boolean.

>     Controls whether Plone's commenting system is enabled for this content type.

Reviewer 2: ```suggestion
    Controls whether Plone's commenting system is enabled by default for this content type.
```
It also needs the behavior to be enabled AFAIK.

    Usually the same as the content type name.

add_permission
    Id of the permission controlling whether the current user has permission to add this content type.

klass
    Dotted path to the Python class for this content type.

schema
    Dotted path to the schema for this content type.

model_file
    Location of an XML file to load as the content type's schema.
    This is an alternative to `schema` and `model_source`.

model_source
    Inline XML schema for the content type.
    This is an alternative to `schema` and `model_file`.

behaviors

>     List of {ref}`behaviors <backend-behaviors-label>` enabled for this content type.

Reviewer: ```suggestion
    List of {doc}`behaviors <behaviors>` enabled for this content type.
```

global_allow
    Boolean.

>     Set to True to allow adding the content type anywhere in the site where the user has permission.

Reviewer: ```suggestion
    Set to `True` to allow adding the content type anywhere in the site where the user has permission.
```

>     Set to False to only allow adding it inside other content types that include this one in `allowed_content_types`.

Reviewer: ```suggestion
    Set to `False` to only allow adding it inside other content types that include this one in `allowed_content_types`.
```

filter_content_types

>     Boolean. Controls which content types can be added inside this one.

Reviewer: ```suggestion
:   Boolean.
    Controls which content types can be added inside this one.
```

>     If True, allow only the types listed in `allowed_content_types`.

Reviewer: ```suggestion
    If `True`, allow only the types listed in `allowed_content_types`.
```

>     If False, allow any content type that the user has permission to add.

Reviewer: ```suggestion
    If `False`, allow any content type that the user has permission to add.
```

allowed_content_types
    List of content types which can be added inside this one.
    Only used if `filter_content_types` is True.

allow_discussion

>     Boolean. Controls whether Plone's commenting system is enabled for this content type.

Reviewer: ```suggestion
:   Boolean.
    Controls whether Plone's commenting system is enabled for this content type.
```

> :   Dotted path to the schema for this content type.

Reviewer 2: Shall we mention, one of those must be set. Otherwise it fails silently. Event if only behaviors are used and no direct fields are needed.

Author: @jensens I updated it to clarify.

`action` elements

>     Defines additional {ref}`actions <backend-portal-actions-label>` which are available for this content type.

Reviewer: ```suggestion
    Defines additional {doc}`actions <portal-actions>` which are available for this content type.
```

```

The following FTI properties are only used in the Classic UI:

> ```{glossary}

Reviewer: Convert this glossary to a definition list, same as the previous.

:sorted: true
icon_expr
    TALES expression returning the name of one of the registered icons.

>     See {ref}`classic-ui-icons-label`

Reviewer: ```suggestion
    See {doc}`icons`.
```

link_target

>     TODO

Reviewer: ```suggestion
    ```{todo} add definition of `link_target`.
    ```
```

`default_view`
:   Name of the default view used to display this content type.

`default_view_fallback`

> :   ```{todo} add definition of `default_view_fallback `.

Reviewer: The todos are still wonky. Let's try putting the content on a separate line for both.

```suggestion
:   ```{todo}
    Add definition of `default_view_fallback`.
```

Reviewer: It looks like [MyST still chokes on the `todo` directive as the first line in a definition](https://deploy-preview-1498--6-docs-plone-org.netlify.app/backend/content-types/index.html).

![Screen Shot 2023-05-15 at 10 14 02 AM](https://github.com/plone/documentation/assets/102112/576ed3be-def3-4386-b8e0-d32bd83dced2)

Let's keep it simple and use plain old text, and remove the directives and backtick fences.

And I'm clearly lacking ☕ . With any TODO, it is a good idea to create an issue and reference the issue in the TODO to encourage and facilitate contributions.

```suggestion
:   TODO: add definition of `default_view_fallback `.
    See [link to issue].
```

Author: @stevepiercy I did some more research, added a definition for default_view_fallback, and removed link_target because it appears to be unused and not required.

    Name of the default view used to display this content type.

default_view_fallback

>     TODO

Reviewer: ```suggestion
    ```{todo} add definition of `default_view_fallback `.
    ```
```

> :   TALES expression returning the name of one of the registered icons.

Reviewer 2: btw, do we have a glossary term defined for TALES?

Author: @jensens yes, now linked.

@stevepiercy Is there any way to make Sphinx show glossary terms as an overlay on hover/click rather than linking to the glossary? It would be handy to see them without the overhead of navigating to a separate page and back. But it's a nice-to-have, not essential.

Reviewer: Yes, it has been done in various ways, but there would need to be some hacking. See https://github.com/plone/training/issues/540 for a possible solution.
